File: WCCBO2/Shizuku2Lib/DateTimeCommunicator.cs
using System.IO.BACnet;

namespace Shizuku2.BACnet
{
  /// <summary>Shizuku2のDateTimeControllerとの通信ユーティリティクラス</summary>
  public class DateTimeCommunicator : PresentValueReadWriter
  {

    #region 定数宣言

    /// <summary>Device ID</summary>
    public const uint DATETIME_CONTROLLER_DEVICE_ID = 1;

    /// <summary>排他的ポート番号</summary>
    public const int DATETIME_CONTROLLER_EXCLUSIVE_PORT = 0xBAC0 + (int)DATETIME_CONTROLLER_DEVICE_ID;

    /// <summary>Dummy DeviceのBACnetアドレス</summary>
    private readonly BacnetAddress bacAddress;

    #endregion

    #region 列挙型

    /// <summary>項目</summary>
    private enum MemberNumber
    {
      /// <summary>現在のシミュレーション上の日時</summary>
      CurrentDateTimeInSimulation = 1,
      /// <summary>加速度</summary>
      AccelerationRate = 2,
      /// <summary>現実時間の基準日時</summary>
      BaseRealDateTime = 3,
      /// <summary>シミュレーション上の基準日時</summary>
      BaseAcceleratedDateTime = 4,
      /// <summary>シミュレーション上の終了日時</summary>
      EndDateTime = 5,
      /// <summary>計算遅延中か否か</summary>
      IsDelayed = 6,
      /// <summary>計算完了済か否か</summary>
      IsFinished = 7,
      /// <summary>一時停止までの秒数</summary>
      PauseTimer = 8,
      /// <summary>一時停止中か否か</summary>
      IsPaused = 9
    }

    #endregion

    #region コンストラクタ

    /// <summary>インスタンスを初期化する</summary>
    /// <param name="id">通信に使うBACnet DeviceのID</param>
    /// <param name="name">通信に使うBACnet Deviceの名前</param>
    /// <param name="ipAddress">エミュレータのIPアドレス（「xxx.xxx.xxx.xxx」の形式）</param>
    public DateTimeCommunicator(uint id, string name = "anoymous device", string ipAddress = "127.0.0.1")
      : base(id)
    {
      bacAddress = new BacnetAddress(BacnetAddressTypes.IP, ipAddress + ":" + DATETIME_CONTROLLER_EXCLUSIVE_PORT.ToString());
    }

    #endregion

    #region インスタンス変数・プロパティ

    /// <summary>COV登録済みか否か</summary>
    private bool covSubscribed = false;

    /// <summary>現実時間の基準日時</summary>
    private DateTime baseRealDateTime = new DateTime(1999, 1, 1, 0, 0, 0);

    /// <summary>シミュレーション上の基準日時</summary>
    private DateTime baseAcceleratedDateTime = new DateTime(1999, 1, 1, 0, 0, 0);

    /// <summary>加速度[-]</summary>
    private int accelerationRate = 0;

    /// <summary>シミュレーション上の現在日時を取得する</summary>
    /// <remarks>
    /// 正しい値を取得するには予めSyncDateTime()で時刻を同期させるか、SubscribeDateTimeCOV()メソッドでCOV通知を受けるようにしておく必要がある
    /// </remarks>
    public DateTime CurrentDateTime
    {
      get
      {
        //一時停止中は初期日時
        if (IsPaused) return baseAcceleratedDateTime;
        //計算中は日時を計算
        else return baseAcceleratedDateTime.AddSeconds
          ((DateTime.Now - baseRealDateTime).TotalSeconds * accelerationRate);
      }
    }

    /// <summary>一時停止中か否かを取得する</summary>
    public bool IsPaused { get; private set; } = false;

    #endregion

    #region 現在日時取得関連

    /// <summary>シミュレーション日時に関するCOVを登録する</summary>
    /// <returns>登録成功の真偽</returns>
    public bool SubscribeDateTimeCOV()
    {
      if (!covSubscribed)
      {
        //加速度の変更を監視
        BacnetObjectId boID = new BacnetObjectId(BacnetObjectTypes.OBJECT_ANALOG_OUTPUT, (uint)MemberNumber.AccelerationRate);
        bool success = client.SubscribeCOVRequest(bacAddress, boID, (uint)MemberNumber.AccelerationRate, false, false, 3600);
        if (!success) return false;

        //一時停止を監視
        boID = new BacnetObjectId(BacnetObjectTypes.OBJECT_BINARY_INPUT, (uint)MemberNumber.IsPaused);
        success = client.SubscribeCOVRequest(bacAddress, boID, (uint)MemberNumber.IsPaused, false, false, 3600);
        if (!success) return false;

        //COVイベント登録
        covSubscribed = true;
        client.OnCOVNotification += Client_OnCOVNotification;
      }

      //日時を更新
      SyncDateTime();

      return true;
    }

    private void Client_OnCOVNotification(
      BacnetClient sender, BacnetAddress adr, byte invokeId, uint subscriberProcessIdentifier,
      BacnetObjectId initiatingDeviceIdentifier, BacnetObjectId monitoredObjectIdentifier,
      uint timeRemaining, bool needConfirm, ICollection<BacnetPropertyValue> values, BacnetMaxSegments maxSegments)
    {
      //加速度が変化した場合
      bool hasChanged = false;
      UInt16 port = BitConverter.ToUInt16(new byte[] { adr.adr[5], adr.adr[4] });
      if (
        port == DATETIME_CONTROLLER_EXCLUSIVE_PORT &&
        monitoredObjectIdentifier.type == BacnetObjectTypes.OBJECT_ANALOG_OUTPUT &&
        monitoredObjectIdentifier.instance == (uint)MemberNumber.AccelerationRate)
      {
        //変更後の加速度を取得
        foreach (BacnetPropertyValue vl in values)
        {
          if (vl.property.GetPropertyId() == BacnetPropertyIds.PROP_PRESENT_VALUE)
          {
            accelerationRate = (int)((float)vl.value[0].Value);
            break;
          }
        }

        hasChanged = true;
      }

      //一時停止状態が変化した場合
      else if (
        port == DATETIME_CONTROLLER_EXCLUSIVE_PORT &&
        monitoredObjectIdentifier.type == BacnetObjectTypes.OBJECT_BINARY_INPUT &&
        monitoredObjectIdentifier.instance == (uint)MemberNumber.IsPaused)
      {
        foreach (BacnetPropertyValue vl in values)
        {
          if (vl.property.GetPropertyId() == BacnetPropertyIds.PROP_PRESENT_VALUE)
          {
            hasChanged = true;
            IsPaused = (uint)vl.value[0].Value == 1u;
            break;
          }
        }
      }

      //変更があった場合には基準日時を更新
      if (hasChanged)
      {
        //現実時間の基準日時を取得
        for (int i = 0; i < 3; i++)
        {
          baseRealDateTime = getBaseRealDateTime(out bool suc);
          if (suc) break;
          Thread.Sleep(100);
        }

        //加速時間の基準日時を取得
        for (int i = 0; i < 3; i++)
        {
          baseAcceleratedDateTime = getBaseAcceleratedDateTime(out bool suc);
          if (suc) break;
          Thread.Sleep(100);
        }
      }
    }

    /// <summary>日時を同期させる</summary>
    /// <returns>同期が成功したか否か</returns>
    public bool SyncDateTime()
    {
      bool suc;

      //一時停止の情報を取得
      bool isp = GetIsPaused(out suc);
      if (!suc) return false;

      //加速度を取得
      int acc = GetAccelerationRate(out suc);
      if (!suc) return false;

      //現実時間の基準日時を取得
      DateTime bdt = getBaseRealDateTime(out suc);
      if (!suc) return false;

      //加速時間の基準日時を取得
      DateTime adt = getBaseAcceleratedDateTime(out suc);
      if (!suc) return false;

      IsPaused = isp;
      accelerationRate = acc;
      baseRealDateTime = bdt;
      baseAcceleratedDateTime = adt;
      return true;
    }

    #endregion

    #region 加速度関連の処理

    /// <summary>加速度[-]を変える</summary>
    /// <param name="accelerationRate">加速度[-]</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    public void ChangeAccelerationRate
      (int accelerationRate, out bool succeeded)
    {
      WritePresentValue(bacAddress,
        BacnetObjectTypes.OBJECT_ANALOG_OUTPUT,
        (uint)MemberNumber.AccelerationRate,
        new BacnetValue(BacnetApplicationTags.BACNET_APPLICATION_TAG_REAL, (float)accelerationRate),
        out succeeded);
    }

    /// <summary>加速度[-]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>加速度[-]</returns>
    public int GetAccelerationRate(out bool succeeded)
    {
      return (int)ReadPresentValue<float>(bacAddress,
        BacnetObjectTypes.OBJECT_ANALOG_OUTPUT,
        (uint)MemberNumber.AccelerationRate,
        out succeeded);
    }

    /// <summary>現実時間の基準日時を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>現実時間の基準日時</returns>
    private DateTime getBaseRealDateTime(out bool succeeded)
    {
      return ReadPresentValue<DateTime>(bacAddress,
        BacnetObjectTypes.OBJECT_DATETIME_VALUE,
        (uint)MemberNumber.BaseRealDateTime,
        out succeeded);
    }

    /// <summary>シミュレーション上の基準日時を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>シミュレーション上の基準日時</returns>
    private DateTime getBaseAcceleratedDateTime(out bool succeeded)
    {
      return ReadPresentValue<DateTime>(bacAddress,
        BacnetObjectTypes.OBJECT_DATETIME_VALUE,
        (uint)MemberNumber.BaseAcceleratedDateTime,
        out succeeded);
    }

    #endregion

    #region 一時停止関連の処理

    /// <summary>一時停止中か否かを取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>一時停止中か否か</returns>
    public bool GetIsPaused(out bool succeeded)
    {
      return 1 == ReadPresentValue<uint>
        (bacAddress, BacnetObjectTypes.OBJECT_BINARY_INPUT, (uint)MemberNumber.IsPaused, out succeeded);
    }

    /// <summary>一時停止までの秒数[sec]を変える</summary>
    /// <param name="pauseTimer">一時停止までの秒数[sec]</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    public void ChangePauseTimer
      (int pauseTimer, out bool succeeded)
    {
      WritePresentValue(bacAddress,
        BacnetObjectTypes.OBJECT_ANALOG_OUTPUT,
        (uint)MemberNumber.PauseTimer,
        new BacnetValue(BacnetApplicationTags.BACNET_APPLICATION_TAG_REAL, (float)pauseTimer),
        out succeeded);
    }

    /// <summary>一時停止までの秒数[sec]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>一時停止までの秒数[sec]</returns>
    public int GetPauseTimer(out bool succeeded)
    {
      return (int)ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_OUTPUT, (uint)MemberNumber.PauseTimer, out succeeded);
    }

    #endregion

    #region その他の情報取得

    /// <summary>計算を終えるシミュレーション上の日時を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>計算を終えるシミュレーション上の日時</returns>
    public DateTime GetEndDateTime(out bool succeeded)
    {
      return ReadPresentValue<DateTime>(bacAddress,
        BacnetObjectTypes.OBJECT_DATETIME_VALUE,
        (uint)MemberNumber.EndDateTime,
        out succeeded);
    }

    /// <summary>計算が遅延しているか否かを取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>計算が遅延しているか否か</returns>
    public bool GetIsDelayed(out bool succeeded)
    {
      return 1 == ReadPresentValue<uint>
        (bacAddress, BacnetObjectTypes.OBJECT_BINARY_INPUT, (uint)MemberNumber.IsDelayed, out succeeded);
    }

    /// <summary>計算が終了済か否かを取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>計算が終了済か否か</returns>
    public bool GetIsFinished(out bool succeeded)
    {
      return 1 == ReadPresentValue<uint>
        (bacAddress, BacnetObjectTypes.OBJECT_BINARY_INPUT, (uint)MemberNumber.IsFinished, out succeeded);
    }

    #endregion

  }
}

File: Samples/python/bacpypes3/DateTimeCommunicator.py
import datetime
import asyncio
from enum import Enum

from bacpypes3.pdu import Address
from bacpypes3.primitivedata import ObjectIdentifier, Real

import sys,os
sys.path.append(os.path.dirname(__file__))
import PresentValueReadWriter

class DateTimeCommunicator(PresentValueReadWriter.PresentValueReadWriter):
    """Shizuku2のDateTimeControllerとの通信ユーティリティクラス
    """  

    # region 定数宣言

    DATETIMECONTROLLER_DEVICE_ID = 1

    DATETIMECONTROLLER_EXCLUSIVE_PORT = 0xBAC0 + DATETIMECONTROLLER_DEVICE_ID

    # endregion

    # region 列挙型定義

    class _member(Enum):
        # 現在のシミュレーション上の日時
        CurrentDateTimeInSimulation = 1
        # 加速度
        AccelerationRate = 2
        # 現実時間の基準日時
        BaseRealDateTime = 3
        # シミュレーション上の基準日時
        BaseAcceleratedDateTime = 4
        # シミュレーション上の終了日時
        EndDateTime = 5
        # 計算遅延中か否か
        IsDelayed = 6
        # 計算完了済か否か
        IsFinished = 7
        # 一時停止までの秒数
        PauseTimer = 8
        # 一時停止中か否か
        IsPaused = 9

    # endregion

    # region 初期化処理

    def __init__(self, id, name='dtComm', device_ip='127.0.0.1', emulator_ip='127.0.0.1', time_out_sec = 1.0):
        """インスタンスを初期化する

        Args:
            id (int): 通信に使うDeviceのID
            name (str): 通信に使うDeviceの名前
            device_ip (str): 通信に使うDeviceのIP Address（xxx.xxx.xxx.xxx）
            emulator_ip (str): エミュレータのIP Address（xxx.xxx.xxx.xxx）
            time_out_sec (float): タイムアウトまでの時間[sec]
        """
        super().__init__(id, name, device_ip, emulator_ip, time_out_sec)
        self.__target_ip = emulator_ip + ':' + str(self.DATETIMECONTROLLER_EXCLUSIVE_PORT)

        # DateTimeのCOV登録状況
        self.__acccov_task = None
        self.__ispcov_task = None

        # 日時を計算するための加速度・基準日時
        self.__acc_rate = 0
        self.__base_real_datetime = datetime.datetime.today()
        self.__base_sim_datetime = datetime.datetime.today()

        # 一時停止状況
        self.is_paused = False

    # endregion

    # region 現在日時取得関連

    def unsubscribe_date_time_cov(self):
        """シミュレーション日時に関する情報のCOVイベントを解除する
        Args:None
        Returns:None
        """ 
        self.__acccov_task.cancel()
        self.__ispcov_task.cancel()
        self.__acccov_task = None
        self.__ispcov_task = None


    async def subscribe_date_time_cov(self):
        """シミュレーション日時に関する情報のCOVイベントを登録する
        Args:None
        Returns:None
        """
        self.__acccov_task = asyncio.create_task(self.acccov_loop())
        self.__ispcov_task = asyncio.create_task(self.ispcov_loop())


    async def acccov_loop(self):
        # 既に登録されている場合には日時だけ更新して二重登録を回避
        if self.__acccov_task == None:
            return await self.sync_date_time()
        try:
            async with self.bacdevice.change_of_value(
                address=Address(self.__target_ip),
                subscriber_process_identifier=self.id,
                monitored_object_identifier=ObjectIdentifier('analogOutput:' + str(DateTimeCommunicator._member.AccelerationRate.value)), # 加速度
                lifetime=60*60*24*365, #1年間
                issue_confirmed_notifications=True
            ) as scm: #SubscriptionContextManager
                self.__acccov_scribed = True
                while True:
                    property_identifier, property_value = await scm.get_value()
                    if(f"{property_identifier}"=='present-value'):
                        await self.sync_date_time()
        except Exception as err:
            return False
        

    async def ispcov_loop(self):
        # 既に登録されている場合には日時だけ更新して二重登録を回避
        if self.__ispcov_task == None:
            return await self.sync_date_time()
        try:
            async with self.bacdevice.change_of_value(
                address=Address(self.__target_ip),
                subscriber_process_identifier=self.id + 1,
                monitored_object_identifier=ObjectIdentifier('binaryInput:' + str(DateTimeCommunicator._member.IsPaused.value)), # 一時停止
                lifetime=60*60*24*365, #1年間
                issue_confirmed_notifications=True
            ) as scm: #SubscriptionContextManager
                self.__ispcov_scribed = True
                while True:
                    property_identifier, property_value = await scm.get_value()
                    if(f"{property_identifier}"=='present-value'):
                        self.is_paused = (property_value._value == 1)
                        await self.sync_date_time()
        except Exception as err:
            return False
        

    async def sync_date_time(self):
        """日時をエミュレータに同期させる
        Args:
        Returns:
            bool: 成功したか否か
        """ 
        val = await self.get_acceleration_rate()
        if val[0]: self.__acc_rate = val[1]
        else: return False 
        val = await self.read_present_value(self.__target_ip, 'datetimeValue:' + str(DateTimeCommunicator._member.BaseRealDateTime.value))
        if val[0]: self.__base_real_datetime = val[1]
        else: return False
        val = await self.read_present_value(self.__target_ip, 'datetimeValue:' + str(DateTimeCommunicator._member.BaseAcceleratedDateTime.value))
        if val[0]: self.__base_sim_datetime = val[1]
        else: return False
        return True

    def current_date_time(self):
        """現在の日時を取得する
        Args:
        Returns:
            datetime: 現在の日時
        """
        if self.is_paused:
            return self.__base_sim_datetime
        else: 
            return (datetime.datetime.today() - self.__base_real_datetime) * self.__acc_rate + self.__base_sim_datetime

    # endregion

    # region 加速度関連

    async def get_acceleration_rate(self):
        """加速度[-]を取得する

        Returns:
            list: 読み取り成功の真偽,加速度[-]
        """
        return await self.read_present_value(self.__target_ip,'analogOutput:' + str(DateTimeCommunicator._member.AccelerationRate.value))

    async def change_acceleration_rate(self, acceleration_rate):
        """加速度[-]を変える
        Args:
            acceleration_rate (float): 加速度[-]
        Returns:
            bool:命令が成功したか否か
            """
        return await self.write_present_value(self.__target_ip,'analogOutput:' + str(DateTimeCommunicator._member.AccelerationRate.value),Real(acceleration_rate))

    # endregion

    # region 一時停止関連の処理

    async def get_pause_timer(self):
        """一時停止までの時間[sec]を取得する

        Returns:
            list: 読み取り成功の真偽,一時停止までの時間[sec]
        """
        return await self.read_present_value(self.__target_ip,'analogOutput:' + str(DateTimeCommunicator._member.PauseTimer.value))

    async def change_pause_timer(self, pause_timer):
        """一時停止までの時間[sec]を変える
        Args:
            pause_timer (float): 一時停止までの時間[sec]
        Returns:
            bool:命令が成功したか否か
            """
        return await self.write_present_value(self.__target_ip,'analogOutput:' + str(DateTimeCommunicator._member.PauseTimer.value),Real(pause_timer))

    async def get_is_paused(self):
        """計算が一時停止中か否かを取得する

        Returns:
            list: 読み取り成功の真偽,計算が一時停止中か否か
        """
        val = await self.read_present_value(self.__target_ip,'binary-input:' + str(DateTimeCommunicator._member.IsPaused.value))
        return val[0], (val[1] == 1)

    # endregion

    # region その他の処理

    async def get_end_dateTime(self):
        """計算を終えるシミュレーション上の日時を取得する

        Returns:
            list: 読み取り成功の真偽,計算を終えるシミュレーション上の日時
        """
        return await self.read_present_value(self.__target_ip,'datetime-value:' + str(DateTimeCommunicator._member.EndDateTime.value))


    async def get_is_delayed(self):
        """計算遅延の有無を取得するを取得する

        Returns:
            list: 読み取り成功の真偽,計算遅延の有無を取得する
        """
        val = await self.read_present_value(self.__target_ip,'binary-input:' + str(DateTimeCommunicator._member.IsDelayed.value))
        return val[0], (val[1] == 1)
    

    async def get_is_finished(self):
        """計算が終了済か否かを取得する

        Returns:
            list: 読み取り成功の真偽,計算が終了済か否か
        """
        val = await self.read_present_value(self.__target_ip,'binary-input:' + str(DateTimeCommunicator._member.IsFinished.value))
        return val[0], (val[1] == 1)

    # endregion

# region サンプル

async def main():
    pv_rw = DateTimeCommunicator(
        id=50,
        name= 'myDevice',
        device_ip='127.0.0.1/255.255.255.0',
        emulator_ip='127.0.0.1',
        time_out_sec=1
    )

    # 日時のCOVを登録
    print('Subscribe COV')
    await pv_rw.subscribe_date_time_cov()
    await pv_rw.sync_date_time()

    # 無限ループで日時を表示
    while True:
        print(pv_rw.current_date_time().strftime('%Y/%m/%d %H:%M:%S'))
        val = await pv_rw.get_is_delayed()
        print('Is delayed? ' + (str(val[1]) if val[0] else ' 通信失敗'))
        val = await pv_rw.get_is_finished()
        print('Is finished? ' + (str(val[1]) if val[0] else ' 通信失敗'))
        val = await pv_rw.get_is_paused()
        print('Is paused? ' + (str(val[1]) if val[0] else ' 通信失敗'))
        print()
        await asyncio.sleep(1)
        pass

if __name__ == "__main__":
    asyncio.run(main())

# endregion